core/arch/arm/plat-sam/freq.c

```diff
@@ -12,8 +12,102 @@
 #include <kernel/panic.h>
 #include <libfdt.h>
 
+#define OPP_RATES_MAX 8
+
+/* the clock and its rates for CPU OPP */
+struct clk_rates {
+	struct clk *clk;
+	size_t rate_num; /* the number of valid clock rates in @rates */
+	unsigned long rates[OPP_RATES_MAX];
+};
+
+static struct clk_rates *opp_rates;
 static unsigned long freq;
 
+static TEE_Result get_rates_array(struct clk *clk, size_t start_index,
+				  unsigned long *rates, size_t *nb_elts)
+{
+	if (!opp_rates)
+		panic("Invalid CPU OPP Rates Array");
+
+	if (opp_rates->clk != clk)
+		panic("Invalid clock");
+
+	if (!rates) {
+		*nb_elts = opp_rates->rate_num;
+
+		return TEE_SUCCESS;
+	}
+
+	if (start_index > opp_rates->rate_num ||
+	    start_index + *nb_elts > opp_rates->rate_num) {
+		EMSG("Bad parameter(s): start_index %d, nb_elts %d",
+		     start_index, *nb_elts);
+
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memcpy(rates, &opp_rates->rates[start_index],
+	       *nb_elts * sizeof(*rates));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result dt_get_opp_hz(const void *fdt, int node, unsigned long *value)
+{
+	const char *property = "opp-hz";
+	const fdt64_t *p = NULL;
+	int len = 0;
+
+	p = fdt_getprop(fdt, node, property, &len);
+	if (!p)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len != sizeof(*p))
+		return TEE_ERROR_BAD_FORMAT;
+
+	*value = fdt64_ld(p);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result opp_rates_setup(const void *fdt, int node, struct clk *clk)
+{
+	const char *compatible = "operating-points-v2";
+	const struct fdt_property *prop = NULL;
+	unsigned long rate = 0;
+	int opp_table = 0;
+	int offset = 0;
+	int len = 0;
+
+	prop = fdt_get_property(fdt, node, compatible, &len);
+	if (!prop)
+		return TEE_SUCCESS;
+
+	opp_rates = calloc(1, sizeof(*opp_rates));
+	if (!opp_rates) {
+		EMSG("Fail to alloc opp_rates");
+
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+	memset(opp_rates, 0, sizeof(*opp_rates));
+
+	offset = (int)prop - (int)fdt;
+	if (fdt_node_check_compatible(fdt, offset, compatible))
```

Comment:
I don't really get the above operation, isn't the node you're searching in simply "node"?
Why is `if (fdt_node_check_compatible(fdt, node, compatible))` not enough?

Reply:
Here the node in search is [opp-table](https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/dts/sama7g5.dtsi#L40) and the compatible value is `operating-points-v2`. The `node` parameter of the function is for the [`cpu node`](https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/dts/sama7g5.dtsi#L29) in which `cpu_opp_table` is one of its items.

Reply:
prefer
```patch
- 	offset = (int)prop - (int)fdt;
+	offset = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(*cuint));
+	if (offset < 0) {
+		// error case
+	}
+
 	if (fdt_node_check_compatible(fdt, offset, compatible)) {
 		panic("Invalid opp-table");

Reply:
updated with the code at [`core/drivers/clk/sam/at91_cpu_opp.c:96`](https://github.com/OP-TEE/optee_os/pull/6885/files#diff-501b0efc0f38b2e470cc29c13962cd636701f80226260de47265cf185e6b8be3R96), thanks.